pkg/progress/action.go

```diff
@@ -0,0 +1,215 @@
+) (map[string]float64, float64, error) {
+	var (
+		phaseWeights = map[string]float64{}
+		totalWeight  = 0.0
+	)
+
+	actionSet, err := client.CrV1alpha1().ActionSets(namespace).Get(ctx, actionSetName, metav1.GetOptions{})
+	if err != nil {
+		return nil, 0.0, err
+	}
+
+	for _, action := range actionSet.Spec.Actions {
+		blueprintName := action.Blueprint
+		blueprint, err := client.CrV1alpha1().Blueprints(actionSet.GetNamespace()).Get(ctx, blueprintName, metav1.GetOptions{})
+		if err != nil {
+			return nil, 0.0, err
+		}
+
+		if err := validate.Blueprint(blueprint); err != nil {
+			return nil, 0.0, err
+		}
+
+		blueprintAction, exists := blueprint.Actions[action.Name]
+		if !exists {
+			return nil, 0.0, fmt.Errorf("missing blueprint action: %s", action.Name)
+		}
+
+		for _, phase := range blueprintAction.Phases {
+			phaseWeight := weight(&phase)
+			phaseWeights[phase.Name] = phaseWeight
+			totalWeight += phaseWeight
+		}
+	}
+
+	return phaseWeights, totalWeight, nil
+}
+
+func updateActionsProgress(
+	ctx context.Context,
+	client versioned.Interface,
+	actionSet *crv1alpha1.ActionSet,
+	phaseWeights map[string]float64,
+	totalWeight float64,
+	now time.Time,
+) error {
+	if err := validate.ActionSet(actionSet); err != nil {
+		return err
+	}
+
+	// assess the state of the phases in all the actions to determine progress
+	currentWeight := 0.0
+	for _, action := range actionSet.Status.Actions {
+		for _, phase := range action.Phases {
+			if phase.State != crv1alpha1.StateComplete {
+				continue
+			}
+			currentWeight += phaseWeights[phase.Name]
+		}
+	}
+
+	percent := (currentWeight / totalWeight) * 100.0
+	progressPercent := strconv.FormatFloat(percent, 'f', 2, 64)
+	if progressPercent == progressPercentNotStarted {
+		progressPercent = progressPercentStarted
+	}
+
+	fields := field.M{
+		"actionSet": actionSet.GetName(),
+		"namespace": actionSet.GetNamespace(),
+		"progress":  progressPercent,
+	}
+	log.Debug().Print("updating action progress", fields)
+
+	return updateActionSet(ctx, client, actionSet, progressPercent, now)
+}
+
+func weight(phase *crv1alpha1.BlueprintPhase) float64 {
+	if longRunning(phase) {
+		return weightHeavy
+	}
+	return weightNormal
+}
+
+func longRunning(phase *crv1alpha1.BlueprintPhase) bool {
+	for _, f := range longRunningFuncs {
+		if phase.Func == f {
+			return true
+		}
+	}
+
+	return false
+}
+
+func updateActionSet(
+	ctx context.Context,
+	client versioned.Interface,
+	actionSet *crv1alpha1.ActionSet,
+	progressPercent string,
+	lastTransitionTime time.Time,
+) error {
+	updateFunc := func(actionSet *crv1alpha1.ActionSet) error {
+		metav1Time := metav1.NewTime(lastTransitionTime)
+
+		actionSet.Status.Progress.PercentCompleted = progressPercent
+		actionSet.Status.Progress.LastTransitionTime = &metav1Time
+		return nil
+	}
+
+	if err := reconcile.ActionSet(ctx, client.CrV1alpha1(), actionSet.GetNamespace(), actionSet.GetName(), updateFunc); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
```suggestion
	return reconcile.ActionSet(ctx, client.CrV1alpha1(), actionSet.GetNamespace(), actionSet.GetName(), updateFunc)
```
error check can be skipped